docs/rfc/rfc-010-p2p-light-client.rst

```diff
@@ -0,0 +1,127 @@
+relayer, it's maintainers, and operators) who would be able to take advantage
+of p2p light client, before switching to libp2p. To our knowledge there are no
+external implementations of this p2p protocol, and it seems unlikely that a
+team would implement this directly ahead of its impending removal.
+
+User Cases
+~~~~~~~~~~
+
+This RFC makes a few assumptions about the use cases and users of light
+clients in tendermint.
+
+One of the most active use cases for light clients is in the implementation of
+the IBC relayer. Thus, we expect that providing P2P light clients might
+increase the reliability of relayers and reduce the cost of running a relayer,
+because relayer operators won't have to decide between rely on public RPC
+endpoints (unreliable) or running their own full nodes (expensive.) This
+also assumes that there are *no* other uses of the RPC in the relayer, and
+unless the relayers have the option of dropping all RPC use, it's unclear if a
+P2P light client will actually be able to successfully remove the dependency
+on the RPC system.
+
+Given that the primary relayer implementation, is hermes (rust,) it might be
+safe to deliver a version of tendermint that adds a light client rector in
+the full nodes, but that does not provide an implementation of a Go light
+client. This either means that the rust implementation would need support for
+the legacy P2P connection protocol or wait for the libp2p implementation.
+
+Client side light client (e.g. wallets, etc.) users may always want to use (a
+subset) of the RPC rather than connect to the P2P network for an ephemeral
+use.
+
+Discussion
+----------
+
+Implementation Questions
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Most of the complication in the is how to have a long lived light client node
+that *only* runs the light client reactor, as this raises a few questions:
+
+- would users specify a single P2P node to connect to when creating a light
+  client or would they also need/want to discover peers?
+
+- how do we prevent full nodes and validators from allowing their peer slots,
+  which are typically limited, from filling with light clients? If
+  light-clients aren't limited, how do we prevent light clients from consuming
+  resources on consensus nodes?
+
+- when a light client disconnects from its peers will it need to reset its
+  internal state? does this change if it connects to the same peers?
+
+These issues are primarily present given that the current peer management later
+does not have a particularly good service discovery mechanism nor does it have
+a very sophisticated way of identifying nodes of different types or modes.
+
+Report Evidence
+~~~~~~~~~~~~~~~
+
+The current light client implementation currently has the ability to report
+observed evidence. Either the notional light client reactor needs to be able
+to handle these kinds of requests *or* all light client nodes need to also run
+the evidence reactor. This could be configured at runtime. 
```

Comment:
This is a great point. The light client isn't able to verify most evidence but it can still generate as well as relay evidence to the network. There was also previous talk about designing a protocol such that light clients could listen out for possible attacks and freeze (like IBC does). So it will need to have some form of evidence reactor but similarly to the light reactor it will be different to the full node implementation

Reply:
If you had the full node's evidence reactor (in go; optionally) you could submit evidence, and that doesn't seem like a terrible idea?

Reply:
Yup. I think part of having a p2p implementation would consist of submitting evidence to full nodes via the evidence channel